frontend/src/components/Table/Table.tsx

```diff
@@ -2,63 +2,112 @@
 //
 // SPDX-License-Identifier: AGPL-3.0-or-later
 
-import React, { useState } from "react";
-import { Column, useTable } from "react-table";
+import React, { useEffect, useState } from "react";
+import { Column, useSortBy, useTable } from "react-table";
 import { v4 as uuidv4 } from "uuid";
 import "./Table.scss";
+import { isRecoilValue, RecoilValueReadOnly, useRecoilValueLoadable } from "recoil";
 import cn from "../../utils/bemUtils";
+import ArrowSvg from "../../assets/svg/arrow.svg";
+import { Spinner } from "../Spinner/Spinner";
+import { SortParamsApi } from "../../api/EdnaApi";
 
 interface ColumnExtra {
   manualCellRendering?: boolean;
 }
 
 // eslint-disable-next-line @typescript-eslint/ban-types
 interface LibraryTableProps<T extends object> {
-  data: T[];
   columns: Column<T>[];
   columnExtras?: { [id: string]: ColumnExtra };
   className?: string;
   small?: boolean;
   collapsible?: (x: T) => React.ReactNode;
+  defaultSortedColumn: string;
+  dataOrQuery:
+    | RecoilValueReadOnly<ReadonlyArray<T>>
+    | ((param: SortParamsApi) => RecoilValueReadOnly<T[]>);
 }
 
 // eslint-disable-next-line @typescript-eslint/ban-types
 export function Table<T extends object>({
-  data,
   columns,
   columnExtras,
   className,
   small,
   collapsible,
+  dataOrQuery,
+  defaultSortedColumn,
 }: LibraryTableProps<T>): React.ReactElement {
-  const { getTableProps, getTableBodyProps, headerGroups, rows, prepareRow } = useTable({
-    columns,
-    data,
-  });
+  const isConstant = isRecoilValue(dataOrQuery);
+  const [data, setData] = useState<T[]>([]);
+
+  const {
+    getTableProps,
+    getTableBodyProps,
+    headerGroups,
+    rows,
+    prepareRow,
+    state: { sortBy },
+  } = useTable<T>(
+    {
+      columns,
+      data,
+      initialState: {
+        sortBy: [
+          {
+            id: defaultSortedColumn,
+          },
+        ],
+      },
+      disableMultiSort: true,
+      manualSortBy: !isConstant,
+      disableSortRemove: true,
+    },
+    useSortBy
+  );
+
+  const loadableData = useRecoilValueLoadable(
+    typeof dataOrQuery === "function"
+      ? dataOrQuery(sortBy.length === 0 ? {} : { sortby: sortBy[0].id, desc: sortBy[0].desc })
+      : dataOrQuery
+  );
+
+  useEffect(() => {
+    if (loadableData.state === "hasValue") {
+      setData(loadableData.contents.concat([]));
+    } else {
+      setData([]);
+    }
+  }, [loadableData]);
```

Comment:
Do you need it? Looks likes you use data only in two places and it will be more pretty to just use this check in that places

Reply:
if you say me how to eliminate cyclic dependency in this case (table depends on `data`, `loadableData` depends on table state `sortby`, `data` depends on `loadableData`), I will move this check in that specific space :putya: